design/splitfrr-proposal.md

```diff
@@ -0,0 +1,465 @@
+  PrefixesIPV6 []string `json:"prefixesIpV6,omitempty"`
+}
+
+type Neighbor struct {
+  Address string `json:"address"`
+  Port uint16 `json:"port,omitempty"`
+  PasswordSecret v1.SecretReference `json:"password,omitempty"`
+  HoldTime metav1.Duration `json:"holdTime,omitempty"`
+  KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+  EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+  BFDProfile string `json:"bfdProfile,omitempty"`
+  ToAdvertise Advertise `json:"toAdvertise,omitempty"`
+  ToReceive Receive `json:"toReceive,omitempty"`
+}
+
+type Advertise struct {
+  AllowedPrefixes `json:"allowed,omitempty"`
+  PrefixesWithLocalPref []LocalPrefPrefixes `json:"withLocalPref,omitempty"`
+  PrefixesWithCommunity []CommunityPrefixes `json:"withCommunity,omitempty"`
+}
+
+type Receive struct {
+  AllowedPrefixes `json:"allowed,omitempty"`
+}
+
+type AllowedPrefixes struct {
+  Prefixes []string `json:"prefixes,omitempty"`
+  AllowAll bool     `json:"allowAll,omitempty"`
+}
+
+type LocalPrefPrefixes struct {
+  Prefixes  []string `json:"prefixes,omitempty"`
+  LocalPref int      `json:"localPref,omitempty"`
+}
+
+type CommunityPrefixes struct {
+  Prefixes  []string `json:"prefixes,omitempty"`
+  Community string   `json:"community,omitempty"`
+}
+
+type BFDProfile struct {
+  Name string `json:"name"`
+  ReceiveInterval *uint32 `json:"receiveInterval,omitempty"`
+  TransmitInterval *uint32 `json:"transmitInterval,omitempty"`
+  DetectMultiplier *uint32 `json:"detectMultiplier,omitempty"`
+  EchoInterval *uint32 `json:"echoInterval,omitempty"`
+  EchoMode *bool `json:"echoMode,omitempty"`
+  PassiveMode *bool `json:"passiveMode,omitempty"`
+  MinimumTTL *uint32 `json:"minimumTtl,omitempty"`
+}
+
+```
+
+This doesn't declare what fields are optional and what are not, but is a good
+approximation of the API.
+
+### Merging multiple configurations applied to the same node
+
+Multiple actors may add configurations selecting the same node.
+There are configurations that may conflict between themselves, leading to errors.
+
+
+This include for example:
+
+- different ASN for the same router (in the same VRF)
+- different ASN for the same neighbor (with the same ip / port)
+- multiple BFD profiles with the same name but different values
+
+When the daemon finds an invalid configuration state of a given node, it will report
+the configuration as invalid and it will clean the FRR configuration.
+
+Incompatibilities aside, merging is straightforward because all the elements
+of the configuration are incremental and allow, for example, to:
+
+- Extend the set of ips we want to advertise to a given neighbor
+Adding an extra neighbor with its set of IPs
+- Extend the set of ips we want to associate a given community to
+- Allow incoming routes for a given neighbor
+
+### Examples
+
+#### Configuration generated by MetalLB
+
+Each MetalLB speaker will generate the configuration related to the node its running on.
+
+For example:
+
```

Comment:
can you add the metallb objects that will lead to that config? (neighbors,adv,...?)